content/algorithms/Dijkstra/Dijkstra.md

```diff
@@ -0,0 +1,241 @@
+                                    ('DUB', 'AMS', 2), ('ZRH', 'GVA', 3), ('GVA', 'LCY', 1)])
+
+pos= nx.planar_layout(flight_path)
+
+# drawing customised nodes
+nx.draw(flight_path, pos, with_labels=True, node_size=1300, node_color='maroon', font_color='white')
+
+# adding edge labels
+nx.draw_networkx_edge_labels(flight_path, pos, edge_labels = nx.get_edge_attributes(flight_path, 'weight'));
+```
+
+## Dijkstra's Algorithm
+
+Dijkstra's algorithm is used to find the shortest path between nodes or commonly from one source node to every other node in the graph, where edge weight represents the cost/ distance between 2 nodes in the case of a weighted graph. It can work with both directed and undirected graphs, but <b>it is not suitable for graphs with NEGATIVE edges.</b><br>
+Time complexity of Dijkstra's algorithm is $O(\ V^{2})$, but with minimum priority queue, it comes down to $O(\ V + E\text{ log } V\ )$
+
+### Algorithm
+
+1. Convert your problem into a graph equivalent.
+2. Create a list of unvisited vertices. 
+3. Assign the starting point as source node with <b>distance(cost)= 0</b> and other node's distance as infinity.
+4. For every unvisited neighbour, calculate the minimum distance from the current node.
+5. The new distance is calculated as `minimum(current distance, distance of previous node + edge weight)`
+6. When all the neighbours have been visited, remove the node from the unvisited list and select the next node with the minimum distance.
+7. Repeat from step 4.
+8. The final graph will represent all the nodes with minimum distance and the algorithm will end.
+
+
+Let's look at the example of the directed graph mentioned above. But, before moving forward, here are some things one should keep in mind. In the following graphs, edge weight defines the distance between 2 nodes, black edge represents unvisited edges, red represents edges that are being traversed, and green represents visited edges. Let's begin!!
+
+
+According to Dijkstra's algorithm,
+- First, assign all stops(nodes) infinite values except the source node (DEL in this case as the path starts from Delhi), which is assigned a value of 0. This is because the distance one needs to cover to reach other nodes is assumed to be unknown and, hence maximum value possible is being assigned. (fig. 1)
+- Dijkstra is based on the greedy approach, which means one needs to select the node with the minimum distance, which is DEL having a distance of 0 units, and this approach is being followed in the whole process.
+- The next step is to start traversing the neighbours of DEL and update the distance of all neighbouring nodes as shown in fig. 2. While updating the distance, always keep in mind that the updated distance should be `minimum(current distance, distance of previous node + edge weight)`. Like, 
+    - DUB : `min(infinity, 7) = 7`
+    - FRA : `min(infinity, 6) = 6`
+    - ZRH : `min(infinity, 5) = 5`
+
+
+![Figure 1&2](Graphs/figure1_2.png "Step 1 and Step 2")
```

Comment:
How were these images generated? It's usually a good idea to keep the code handy to reproduce the figures :)

Reply:
Yes, these images should be generated directly by code in the notebook - showing how to make high-quality visualizations of graph/network data is one of the primary goals of these tutorials!

Reply:
I used [diagrams.net](https://www.diagrams.net/) for the graphs for 2 reasons:
1. I think inserting code for graphs might break the flow of the learner as it happened to me in Dintz and I have like 6 graphs.
2. If you'll look in the images, I can add edge labels, node labels and colour edges, but, I don't know how to add the distance of the nodes in the graph (like infinite initially and then keep updating).

But, yes you're right. I'll try to figure it out and create an easy and understandable graph using networkx.